### ISSUE TITLE
Autoscale VMs do not follow ScaleDown Rules after a Node Failure

### ISSUE THREAD

**Issue by btzq** [apache/cloudstack#9336]
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
Autoscale
~~~

##### CLOUDSTACK VERSION

~~~
4.19.0
~~~

##### STEPS TO REPRODUCE

We are actively using Autoscale Groups which have the following scale down rules. 

Autoscale Rule
- Name: ScaleDownPolicy-0
- Duration (in sec) = 60 Seconds
- Quiet Time (in sec) = 30 seconds

Conditions:
- Counter:  VM CPU - Average Percentage
- Operator: Less Than
- Threshold: 35

These rules work well on normal days. But today, we had a node failure, which the node was hosting some autoscaleVMs

As a result, we experienced this issue, which have already been reported:
- https://github.com/apache/cloudstack/issues/9145

In the case where the VMs were not 'Orphaned', and managed to find its way back to the Autoscale Group (or just werent affected), i noticed that the scale down rule did not work anymore. 

I had VMs that:
- Min : 2 Members
- Max : 6 Members
- Available Instances: 6 _<- This should be 2 Instead._

And after 5 minutes, it still did not scale down. All VMs will running. 
Ive verified the CPU Utilisation of all VMs were only 1% consistently. 

In order to have the service resume, i had to:

- Disable the Autoscale Group
- Delete the VMs
- Re-enable the Autoscale VM

After that, the scaledown rule works, and i had the setup below again:

I had VMs that:
- Min : 2 Members
- Max : 6 Members
- Available Instances: 2

~~~
NA
~~~

##### EXPECTED RESULTS

~~~
NA
~~~

##### ACTUAL RESULTS

~~~
In the case where the VMs were not 'Orphaned', and managed to find its way back to the Autoscale Group (or just happened to be in a host that was not affected), i noticed that the scale down rule did not work anymore.
~~~

**Comment by weizhouapache** [comment]
@btzq 
thanks for reporting
can we/you close one of the two issues ?

**Comment by btzq** [comment]
@weizhouapache which issue are you refering to?

I reported 3 issues regarding autoscaling so far, but they seem to be 3 isolated issues?

- https://github.com/apache/cloudstack/issues/9336
- https://github.com/apache/cloudstack/issues/9318
- https://github.com/apache/cloudstack/issues/9145

If you are refering to https://github.com/apache/cloudstack/issues/9145 , then i need to point out that that ticket was reporting an issue where the VM can become orphans. 

In this ticket https://github.com/apache/cloudstack/issues/9336 , this ticket reports that in either scenarios (including situations where Orphan scenario doesnt take place, the scale down rule doesnt work anymore.

**Comment by weizhouapache** [comment]
> @weizhouapache which issue are you refering to?
> 
> I reported 3 issues regarding autoscaling so far, but they seem to be 3 isolated issues?
> 
> * [Autoscale VMs do not follow ScaleDown Rules after a Node Failure #9336](https://github.com/apache/cloudstack/issues/9336)
> * [Autoscale Group creates Infinite Number Of VMs when unable to startup #9318](https://github.com/apache/cloudstack/issues/9318)
> * [Autoscale Load Balancing Rule looses VMs during a node failure. #9145](https://github.com/apache/cloudstack/issues/9145)
> 
> If you are refering to #9145 , then i need to point out that that ticket was reporting an issue where the VM can become orphans.
> 
> In this ticket #9336 , this ticket reports that in either scenarios (including situations where Orphan scenario doesnt take place), the scale down rule doesnt work anymore.

thanks for the explanation @btzq 

we need to investigate them

**Comment by btzq** [comment]
Hi @DaanHoogland @weizhouapache , we just encountered this issue again after one of our compute servers died.... Please dont put this into deprecated cause it is a very real issue!

**Comment by DaanHoogland** [comment]
@btzq , a new githubfeature is sub-issue so you can group related issues ;) hope you are willing to try it.
#9318 has a possible fix , #9574. Can you test that?

I'll see if I can reproduce the other issues

**Comment by DaanHoogland** [comment]
@btzq , I replayed what I think is your scenario:
- two hosts
- an autoscalegroup with
- min 2, max 6, 30 secs for all timers
- upscale on 50% cpu
- downscale on 35% cpu
using dd zero>null I created load
by doing this on all vms I got to the maximum
I stopped one of the hosts
extra VMs were being created (note that this is strange behaviour as no VMs were disguarded as being part of the group)
I started the host again
the VMs were marked as stopped and the group was now 9 big.
stopping all the dd jobs the load got less and VMS were being stopped and expunged
starting the stopped VMs these were also expunged one by one normalising the situation.

I will investigate some more, but I think it works (more or less) as expected. Can you point to any differences in my scenario with respect to your situation?

**Comment by btzq** [comment]
Hey @DaanHoogland , 

More or less similar scenario, but few notes:

**"extra VMs were being created (note that this is strange behaviour as no VMs were disguarded as being part of the group)"**
- This looks like an issue as well. I think CS should only decide to scale up, if it manages to correctly detect that these VMs are no longer contactable. 
- There is also a scenario where if CS is not able to detect that the VMs are down, the LB might still send traffic to VMs that may not exist. 

**starting the stopped VMs these were also expunged one by one normalising the situation.**
- In my situation, the VMs that were reconnected under the LB/ASG Group was not down. The host may have disconnected for a minute and reconnected again. Our VMs were still running. 
- But if we were to play out your scenario, where the VMs were stopped, then yes, i think it is correct to have the VMs automatically scale down on its own once its started. As long as CS recognizes these VMs as still under the ASG Group.

**Comment by DaanHoogland** [comment]
> [@DaanHoogland](https://github.com/DaanHoogland) For [#9318](https://github.com/apache/cloudstack/issues/9318) and [#9574](https://github.com/apache/cloudstack/pull/9574), they look like the same issue to me.

I don't understand what you mean by this sentence, as one is the solution to the other.

To be brutally honest #9318 seems an operational issue to me, but it still might be solved by #9574. 

> In our case, this issue happened mainly because of storage reasons, where the storage (or corrupted template) cause the VM to be unable to start, so CS tries again and again and again.

Is this about this issue or ablout 9318? I am trying to simulate this one by disconnecting the network between the host and the MS to make sure the VMs aren't stopped.

> Im not sure how to re-simulate this sucessfully again... and also, we dont know how to compile Cloudstack from source... Im a non-technical person actually. But i have a good understanding of how cloudstack works spiritually.

mind over matter ...

**Comment by btzq** [comment]
> Is this about this issue or ablout 9318? I am trying to simulate this one by disconnecting the network between the host and the MS to make sure the VMs aren't stopped.

Yes, that is refering to https://github.com/apache/cloudstack/issues/9318

> I don't understand what you mean by this sentence, as one is the solution to the other.

I mean, they both are related tickets. Yes, i do think https://github.com/apache/cloudstack/issues/9318 can be resolved by https://github.com/apache/cloudstack/pull/9574

**Comment by btzq** [comment]
Maybe its best to put this on hold and see if #9754 resolves the issue.

Im actually thinking of testing this together in the upcoming 4.19.2 release

**Comment by DaanHoogland** [comment]
ok, (assuming it was a typo and you mean 9574) I will close this. Please reopen or create a new issue if problems persist or more info comes available.

**Comment by btzq** [comment]
Ah sorry, yes it was a typo. Got it, will reopen if encountered again
